runtime/lua/lsp/callbacks.lua

```diff
@@ -0,0 +1,531 @@
+--  TODO: textDocument/onTypeFormatting
+--  textDocument/definition
+--  TODO: textDocument/codeAction
+--  TODO: textDocument/codeLens
+--  TODO: textDocument/documentLink
+--  TODO: textDocument/rename
+--
+--
+--  TODO: codeLens/resolve
+--
+--  TODO: documentLink/resolve
+
+local log = require('lsp.log')
+local util = require('nvim.util')
+local lsp_util = require('lsp.util')
+
+local QuickFix = require('nvim.quickfix_list')
+local LocationList = require('nvim.location_list')
+
+local protocol = require('lsp.protocol')
+local errorCodes = protocol.errorCodes
+
+local handle_completion = require('lsp.handle.completion')
+local handle_workspace = require('lsp.handle.workspace')
+
+local CallbackMapping = setmetatable({}, {})
+local CallbackObject = {}
+
+local method_to_callback_object = function(method, create_new)
+  if type(method) ~= 'string' then
+    return nil
+  end
+
+  if CallbackMapping[method] == nil and create_new then
+    CallbackMapping[method] = CallbackObject.new(method)
+  end
+
+  return CallbackMapping[method]
+end
+
+local call_callbacks = function(callback_list, success, params)
+  local results = {}
+
+  for key, callback in ipairs(callback_list) do
+    results[key] = callback(success, params)
+  end
+
+  return unpack(results)
+end
+
+local call_callbacks_for_method = function(method, success, data, default_only, filetype)
+  local cb = method_to_callback_object(method, false)
+
+  if cb == nil then
+    log.debug('Unsupported method:', method)
+    return
+  end
+
+  return cb(success, data, default_only, filetype)
+end
+
+
+CallbackObject.__index = function(self, key)
+  if CallbackObject[key] ~= nil then
+    return CallbackObject[key]
+  end
+
+  return rawget(self, key)
+end
+
+CallbackObject.__call = function(self, success, data, default_only, filetype)
+  if self.name ~= 'nvim/error_callback' and not success then
+    call_callbacks_for_method('nvim/error_callback', false, data)
+  end
+
+  if util.table.is_empty(self.default) then
+      log.trace('Request: "', self.method, '" had no registered callbacks')
+    return nil
+  end
+
+  local callback_list = self:get_list_of_callbacks(default_only, filetype)
+  local results = { }
+
+  for _, cb in ipairs(callback_list) do
+    local current_result = cb(self, data)
+
+    if current_result ~= nil then
+      table.insert(results, current_result)
+    end
+  end
+
+  return unpack(results)
+end
+
+CallbackObject.new = function(method, default_callback, options)
+  options = options or {}
+
+  local object = setmetatable({
+    method = method,
+    options = setmetatable(options, {
+      __index = function(_, key)
+        error(string.format('Cannot get option (%s) for "%s"', key, method))
+      end,
+
+      __newindex = function(_, key, value)
+        error(string.format('Cannot set option (%s) for "%s" to value [%s]', key, method, value))
+      end,
+    }),
+
+    default = {},
+    generic = {},
+    filetype = {},
```

Comment:
@KillTheMule @bfredl 
Now, we have three callback types, default, generic and filetype.
default is used as default callback reference.
I want to solve the problem that default callbacks is always loaded.
I think we don't call `add_default_callback` that is default callback load method at the lsp modules loaded.
And we provide load those load default callbacks method if user want to load default callbacks.
I cannot decide to have to leave default property or delete default property.
If we delete default property and user load default callbacks, they are registered to generic callback.

How do you think?